File: android/app/src/main/java/com/ahmedkhaled/onroute/model/SavedRoute.kt
package com.ahmedkhaled.onroute.model

import android.content.Context
import com.google.android.gms.maps.model.LatLng
import com.squareup.moshi.JsonClass
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory

@JsonClass(generateAdapter = true)
data class SavedRoute(
    val id: String,
    val name: String,
    val originLat: Double,
    val originLng: Double,
    val originName: String,
    val destinationLat: Double,
    val destinationLng: Double,
    val destinationName: String,
    val defaultCategory: String,
    val createdAt: Long
) {
    val originLatLng: LatLng get() = LatLng(originLat, originLng)
    val destinationLatLng: LatLng get() = LatLng(destinationLat, destinationLng)
}

class SavedRoutesStore(context: Context) {
    private val prefs = context.getSharedPreferences("saved_routes", Context.MODE_PRIVATE)
    private val moshi = Moshi.Builder().addLast(KotlinJsonAdapterFactory()).build()
    private val type = Types.newParameterizedType(List::class.java, SavedRoute::class.java)
    private val adapter = moshi.adapter<List<SavedRoute>>(type)
    private val maxRoutes = 5

    fun load(): List<SavedRoute> {
        val json = prefs.getString("routes", null) ?: return emptyList()
        return try { adapter.fromJson(json) ?: emptyList() } catch (_: Exception) { emptyList() }
    }

    fun save(route: SavedRoute) {
        val routes = load().toMutableList()
        routes.removeAll { it.id == route.id }
        routes.add(0, route)
        val trimmed = routes.take(maxRoutes)
        prefs.edit().putString("routes", adapter.toJson(trimmed)).apply()
    }

    fun delete(route: SavedRoute) {
        val routes = load().toMutableList()
        routes.removeAll { it.id == route.id }
        prefs.edit().putString("routes", adapter.toJson(routes)).apply()
    }
}

File: ios/Detour/Models/SavedRoute.swift
import Foundation
import CoreLocation

struct SavedRoute: Identifiable, Codable, Hashable {
    let id: String
    var name: String
    let originLat: Double
    let originLng: Double
    let originName: String
    let destinationLat: Double
    let destinationLng: Double
    let destinationName: String
    var defaultCategory: String
    let createdAt: Date

    var originCoordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: originLat, longitude: originLng)
    }

    var destinationCoordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: destinationLat, longitude: destinationLng)
    }

    static func create(
        name: String,
        origin: CLLocationCoordinate2D,
        originName: String,
        destination: CLLocationCoordinate2D,
        destinationName: String,
        defaultCategory: String = "coffee"
    ) -> SavedRoute {
        SavedRoute(
            id: UUID().uuidString,
            name: name,
            originLat: origin.latitude,
            originLng: origin.longitude,
            originName: originName,
            destinationLat: destination.latitude,
            destinationLng: destination.longitude,
            destinationName: destinationName,
            defaultCategory: defaultCategory,
            createdAt: Date()
        )
    }
}

final class SavedRoutesStore {
    static let shared = SavedRoutesStore()
    private let key = "savedRoutes"
    private let maxRoutes = 5

    func load() -> [SavedRoute] {
        guard let data = UserDefaults.standard.data(forKey: key),
              let routes = try? JSONDecoder().decode([SavedRoute].self, from: data)
        else { return [] }
        return routes
    }

    func save(_ route: SavedRoute) {
        var routes = load()
        routes.removeAll { $0.id == route.id }
        routes.insert(route, at: 0)
        if routes.count > maxRoutes { routes = Array(routes.prefix(maxRoutes)) }
        if let data = try? JSONEncoder().encode(routes) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }

    func delete(_ route: SavedRoute) {
        var routes = load()
        routes.removeAll { $0.id == route.id }
        if let data = try? JSONEncoder().encode(routes) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }
}
